### PR TITLE
Session.id semantic convention

### PR DESCRIPTION
**Pull request by bryce-b**
Discussed this in the client-side sig. Added a basic definition to start discussion.

### CODE REVIEW

File: docs/client/session.md
Comment (scheler) [inline docs/client/session.md:5]: We should include metrics too even though we are not planning on emitting metrics from Client-side apps. This statement could say a session is a collection of all the telemetry data emitted (Spans, Logs, Events and Metrics).

File: docs/client/session.md
Comment (scheler) [inline docs/client/session.md:13]: We could instead say: Client apps generally do not emit metrics since they represent only one user, but in case metrics are emitted then the following attribute should be applied to it as well.

File: docs/client/session.md
Comment (scheler) [inline docs/client/session.md:17]: Can you add this in a yaml at https://github.com/open-telemetry/semantic-conventions/tree/main/model? Since it's a new top level namespace I guess it could be in a file named session.yaml

File: docs/client/session.md
Comment (scheler) [inline docs/client/session.md:25]: Session would be an SDK construct and would be configured when SDK is setup.  Hence the session configuration parameters could only be changed if SDK is re-initialized.

Session timeout and max session duration are both configurable, with some defaults. Could you please add these 2 in a table so they are highlighted as configurable parameters?

File: docs/client/session.md
Comment (breedx-splk) [reply docs/client/session.md:5]: Agreed, and I think we do probably have metrics use cases in mobile (count of slow frame renders, for instance, or other device metrics like memory usage or free space).

File: docs/client/session.md
Comment (breedx-splk) [reply docs/client/session.md:13]: This brings up a good point -- sessions IDs would always be a high-cardinality dimension on a metric. Something I hadn't considered before. Scope level attributes are an interesting possibility, but perhaps components (meter, tracer) have a lifespan that could outlive a session...which poses a different challenge.

File: docs/client/README.md
Comment (breedx-splk) [inline docs/client/README.md:7]: Suggest adding a sentence to anchor the abstract concept of "client" into the common practical types.

```suggestion
This document defines general semantic conventions for client-side applications. 
The most common types of client applications are web applications (that run in a web browser)
and mobile apps (that run on a phone or tablet).
Client applications are differentiated from server applications in that they service a user rather than requests. This adds unique requirements and needs to client-side semantic conventions. 
```

File: docs/client/session.md
Comment (breedx-splk) [inline docs/client/session.md:24]:
```suggestion
Default session timeout SHOULD be 30 minutes. Session timeout period SHOULD be established when the instrumentation is first configured and MUST NOT be updated in the middle of a session. 
```

File: docs/client/session.md
Comment (martinkuba) [reply docs/client/session.md:5]: Is there a use case for including session.id on metrics, given that the values are typically of high cardinality?

File: docs/client/session.md
Comment (martinkuba) [reply docs/client/session.md:17]: Do we intend `session` to be a namespace, so that other attributes related to session can be added in the future? If not, and this is a single attribute, then I think it should be `session_id` instead.

File: docs/client/session.md
Comment (martinkuba) [inline docs/client/session.md:5]: Should we say that the session.id value is generated by the SDK? I want to make sure that it is not misunderstood as a session supplied by the application (e.g. authenticated user session).

Comment (breedx-splk) [reply docs/client/session.md:5]: Nope, I don't have a use case for including `session.id` on metrics. Seems like a problem.

Comment (scheler) [reply docs/client/session.md:5]: My intention was only to indicate that the metric was emitted in the context of a particular session;  I understand the concept of identifying vs describing attributes of a metric is still under consideration. Regarding the attribute being high cardinality, one option is to document this so that the backends could drop this metric attribute during aggregation.

However, since this looks more confusing than helpful, we can drop the idea of adding Session ID on metrics emitted.

File: docs/client/session.md
Comment (scheler) [reply docs/client/session.md:17]: @martinkuba  what would be the namespace of `session_id`? There's an `enduser` namespace, does it make sense to add under that?

File: docs/client/session.md
Comment (scheler) [reply docs/client/session.md:24]: @breedx-splk it's a bit subtle but Session would be an SDK construct and instrumentations aren't aware of it, so this statement could read 'when the SDK is first configured'.

File: docs/client/session.md
Comment (bryce-b) [reply docs/client/session.md:5]: maybe a provision for metrics as exemplars having a session id?

Comment (scheler) [reply docs/client/session.md:5]: I suggest defining Session as follows:

> Session is defined as a single run of an application by an end user, encompassing all the activities performed by the application and the actions executed by the end user. It has a specific duration and is usually concluded when there is no activity, or it can be explicitly ended after a certain maximum duration.
> 
> Consequently, a Session is represented as a collection of Logs, Events, and Spans emitted by the Client Application throughout the Session's duration. Each Session is assigned a unique identifier, which is included as an attribute in the Logs, Events, and Spans generated during the Session's lifecycle.

Comment (bidetofevil) [reply docs/client/session.md:5]: I think the use case is simply to know that a metric is emitted and associated with a particular session, as @scheler mentions. And I think the use case @breedx-splk brings up is perfect: I want to know how many dropped/slow-render frames there within a given sessions to identify its health (or whatever). We would need this association for that to happen in a reasonable way.

In terms of cardinality, I think the utility is to have the association, not using it as a means to group-by, so I think it's just fine.

Comment (bidetofevil) [reply docs/client/session.md:5]: Not sure if we should be that specific in the spec, as the sessionId can be provided by the application if it already has a notion of session, or the SDK if it doesn't or wants to defer the session lifecycle to the SDK.

File: docs/client/session.md
Comment (bidetofevil) [reply docs/client/session.md:13]: > Client apps generally do not emit metrics since they represent only one user 

I actually don't think that's true. They may be interpreted in a different way, but it's a valid use case to associate metrics (in the generic sense) to both sessions and users.

> (meter, tracer) have a lifespan that could outlive a session

Spans can also out live a session depending on how you define a session and a span. The relationship is a fun many-to-many one :-)

File: docs/client/session.md
Comment (bryce-b) [reply docs/client/session.md:25]: Certainly! I'll try to find an example of how this should look.

File: docs/client/session.md
Comment (bidetofevil) [reply docs/client/session.md:17]: Does it need to be a GUID or just something that has a high cardinality and is expected to be unique across some domain or namespace?

I think an optional domain/namespace column makes sense here as there could be different types of sessions that we'd want to track different (e.g. background, foreground, etc.). Having an optional value allows us to do that without adding extra burden for those that don't care.

File: docs/client/session.md
Comment (bidetofevil) [inline docs/client/session.md:20]: I feel this is too prescriptive for me at the spec level. There are use cases where a session can just be the lifetime of the process (e.g. IoT), and having this requirement seems burdensome

File: docs/client/session.md
Comment (bidetofevil) [inline docs/client/session.md:24]: I think this is too prescriptive too. The console app or IoT use case will definitely blow this out of the water.

File: docs/client/session.md
Comment (bryce-b) [reply docs/client/session.md:5]: @martinkuba I think that's reasonable. The term 'user' in the definition could be misconstrued as an authenticated user, as opposed the general "person interacting with the interface". Do you think your suggestion is enough to waylay this confusion? 

@scheler I think that's a pretty good definition, but I don't think the "single run of an application" is necessarily accurate. I imagine an application may crash and be immediately resumed by the user, and still be one session. I guess it depends on how we define "single run"?  I think "a collection of [signals] associated with a single device or user over a span of time" is more precise wording.  Maybe reword it as:
> Session is defined as the period of time encompassing all activities performed by the application and the actions executed by the end user. The period of time is defined by the initial activity and concluded after a period of inactivity, or explicitly ended after a maximum duration is reached.  
>
> Consequently, a Session is represented as a collection of Logs, Events, and Spans emitted by the Client Application throughout the Session's duration. Each Session is assigned a unique identifier, which is included as an attribute in the Logs, Events, and Spans generated during the Session's lifecycle.

File: docs/client/session.md
Comment (bryce-b) [reply docs/client/session.md:20]: It seems like a session would be irrelevant to IoT if the use-case was to be the lifetime of the process. I would see sessions being useful for IoT if they captured discrete interactions with the devices. We should probably discuss IoT more throughly; I'm not extremely well versed in the potential use-cases.

File: docs/client/session.md
Comment (bryce-b) [reply docs/client/session.md:24]: I think it's necessary to have some kind of maximum session length, just due to hardware constraints of recalling session data from whatever is storing it. It doesn't necessarily need to be 4 hours, this is just a place holder.  Distilling the purpose of a session: providing context to the large pool of data by constraining that data to a relevant period of time for a single user... if IoT devices have unlimited sessions, it sort of defeats the purpose. Maybe IoT doesn't need a session: you can get the same unlimited collection of data by constraining you data view to the device the data is generated from (e.g.: `device.id`)

File: docs/client/session.md
Comment (bidetofevil) [reply docs/client/session.md:5]: > The period of time is defined by the initial activity and concluded after a period of inactivity, or explicitly ended after a maximum duration is reached.

I think even this is too prescriptive. I think removing it but including everything else would be precise enough to be useful

File: docs/client/session.md
Comment (bidetofevil) [reply docs/client/session.md:20]: So even without IoT, there is the use console application use case (like a Point of Sales app like McDonald's or something), where the app is always on and there is long periods where there no activity, but you don't want to time that session out, or just send a ping to prevent a timeout.

File: docs/client/session.md
Comment (bidetofevil) [reply docs/client/session.md:24]: Fair enough. Maybe  we can say there is a max length, but that is determined by the application/SDK depending on its specific use case?

File: docs/client/session.md
Comment (scheler) [reply docs/client/session.md:5]: > I think that's a pretty good definition

I took the help of ChatGPT to frame the sentences :-)

Agree on both the suggestions - removing 'single run', and also not prescribing how a session ends. So from the revised text above, we could remove "The period of time is defined by the initial activity and concluded after a period of inactivity, or explicitly ended after a maximum duration is reached.".

File: docs/client/session.md
Comment (bryce-b) [reply docs/client/session.md:24]: I think that's reasonable. Maybe: 
> Implementers SHOULD implement a maximum session length applicable to their use-case

File: docs/client/session.md
Comment (bryce-b) [reply docs/client/session.md:20]: If no spans, logs, or events are recorded before the timeout period occurs then the next occurrence of one of those signals would trigger a session id refresh, thus starting a new session. Depending on the timeout period, this may end up with a session containing one checkout, or possibly a busy period with multiple checkouts. 

Streaming services may set their timeout to the average length of a movie, or generate a "ping" while a movie is being viewed, but that may not be necessary, as instrumented network requests for buffering would refresh the timeout through the viewing of content.

File: docs/client/session.md
Comment (bryce-b) [reply docs/client/session.md:17]: I wonder if we should introduce a `client` namespace? I worry `enduser` implies it's for authenticated users?

File: docs/client/session.md
Comment (bidetofevil) [reply docs/client/session.md:20]: Yeah, the new session triggering after some period of inactivity is the bit I'm not sure of. Perhaps we can support a value of 0 or something that indicates there's no timeout, so folks have to activity think about it in order to use that?

File: docs/client/session.md
Comment (bidetofevil) [inline docs/client/session.md:13]: I believe this was pointed out somewhere, but should be avoid the term agent when talking about instrumentation in general as not all client instrumentation will be done via agents?

File: docs/client/session.md
Comment (jzwc) [reply docs/client/session.md]: I think that this is not necessary to be in the specification.

Using session identifier with OTel makes sense only for applications, that itself deploy some session (discrete client state) concept in its business logic, and want to use in analytics/telemetry. 

The specs purpose should be, IMHO, to enable associating this internal session with OTel signals. OTel spec should not assume much about the session lifecycle details itself. To be meaningful, the state is reset from time to time as dictated by the app business logic.

Let's analyse the McDonald's POS example. 

One can assume, the app there has some internal state that clears with each customer. From the analytics/telemetry point of view, it makes sense to associate the OTel session with this state. 

The state has its clear start (the POS itself clears the state at some moment as not to fuse the previous customers data with the new customer ones). 

The session on the POS ends from the business-loc point of view when the order is sent to backend systems. There, it can continue in the form of distributed traces, that originate on the POS terminal in the scope of the client session there (the session identifier can be distributed in baggage too, eventually). I.e., there can be session-associated signals traced long after the session itself (in the POS terminal) no longer exists. E.g., till the order is handled to the waiting customer.

What happens on the POS terminal in the idle time? 

First of all, are there some signals during the idle time, anyway? If not, it is not important whether the (previous) session identifier is cached or set somewhere or not, as it is no used. If there are some signals, then the logic step would probably be to clear the session identifier when the business-logic session ends in the POS terminal, and trace the idle-time signals as "session-less".

File: docs/client/README.md
Comment (svrnm) [inline docs/client/README.md:7]: I think it would make sense to also call out some of the non-common types (IoT? Desktop applications??)

File: docs/client/README.md
Comment (svrnm) [inline docs/client/README.md:8]: Based on  [this comment](https://github.com/open-telemetry/semantic-conventions/pull/215#discussion_r1276731735) by @bryce-b, does it make sense to provide a definition of "user"? (_person interacting with the interface_: does this exclude synthetic users?)

File: docs/client/session.md
Comment (svrnm) [inline docs/client/session.md:5]: Can a session be used in a server-side opentelemetry instrumentation as well? Let's say as an end user I interact with a website, trigger some client-side API calls which get tagged with the session_id, can the same session_id be applied to API calls that might be executed on my behave in the same context.

File: docs/client/session.md
Comment (svrnm) [inline docs/client/session.md:20]:
```suggestion
| `client.session_id` | string | A UUID. | `00112233-4455-6677-8899-aabbccddeeff` | Opt-In |
```

As far as I understand it UUID is the more common name and also used more consistently throughout the existing semantic conventions. [For reference](https://datatracker.ietf.org/doc/html/rfc4122):

> This specification defines a Uniform Resource Name namespace for
   UUIDs (Universally Unique IDentifier), also known as GUIDs (Globally
   Unique IDentifier).

Is there a need to be more specific around the version of UUID being used?

File: docs/client/session.md
Comment (svrnm) [reply docs/client/session.md]: > Should we say that the session.id value is generated by the SDK? I want to make sure that it is not misunderstood as a session supplied by the application (e.g. authenticated user session).

+1, there should be a clear statement that the session ID must not be re-used from the application, because this can lead to security issues.

File: docs/general/session.md
Comment (scheler) [inline docs/general/session.md:23]: Could you remove these 2 from here as well?

File: docs/client/session.md
Comment (bryce-b) [reply docs/client/session.md]: This was my intention with adding session as a top level namespace!

**Comment by bryce-b** [comment]
the link checker is failing for an unrelated file :\

**Review by breedx-splk — APPROVED** [review]

**Review by MSNev — APPROVED** [review]

**Comment by bryce-b** [comment]
@open-telemetry/specs-approvers @open-telemetry/specs-semconv-maintainers  This is ready for your review!

File: docs/general/session.md
Comment (yurishkuro) [inline docs/general/session.md:19]: * Why "not supplied by the client"?
* Why does it have to be UUID format?
* How does the client exercise control over the life cycle of a session that reflect the business requirements of the app?

Comment (scheler) [reply docs/general/session.md:19]: The 'not supplied by the client' requires rewording. I guess @bryce-b meant the instrumentations should not add this attribute to the signals, but only a processor at SDK level should be adding it. I don't know if this fact needs to be mentioned here though.

Comment (scheler) [reply docs/general/session.md:19]: Regarding UUID, note that the session id has to be unique in the context of a specific resource. For mobile apps, we said the resource will have a `device.id`; For browser apps, we said we will have `browser.visitor.id` but it's not added to the semantic conventions yet. I think we tend to use UUIDs since it's easier to guarantee uniqueness, otherwise we will need to remember a state for that purpose, and this adds complexity on the client devices. Should we reword the text to say `unique id` in place of `UUID`?

Comment (scheler) [reply docs/general/session.md:19]: Regarding the lifecycle of a session, we want to omit specifying it here. This PR originally started describing it but there are many variations to it from different vendors. Since this id is added using a processor at the SDK level we said each vendor could have their own implementation managing the lifecycle of a session and only thing needed in the spec is how the session identifier is represented in the signals. Let me know if this makes sense.

Comment (yurishkuro) [reply docs/general/session.md:19]: > Should we reword the text to say unique id in place of UUID?

Yes, the description should capture semantic requirements, not implementation details like whether it's UUID or not.

Comment (yurishkuro) [reply docs/general/session.md:19]: > but only a processor at SDK level should be adding it

This again sounds like an implementation detail not relevant to the semantic meaning of `session.id`. If my instrumentation has its own ideas about what constitutes a session ID and what its life cycle is, why should semantic convention prevent me from setting it explicitly and instead force on relying some unspecified vendor implementation?

Comment (scheler) [reply docs/general/session.md:19]: Sounds good. 
@bryce-b how about we trim the text to "A unique id to identify a session".

File: docs/general/session.md
Comment (yurishkuro) [inline docs/general/session.md:7]: Who determines when the session starts and ends? Are there common practices of doing so?

Comment (breedx-splk) [reply docs/general/session.md:7]: Session starts when the mobile app is launched or when a web page is loaded. Session typically ends when app is closed or a timeout is reached.

**Comment by breedx-splk** [comment]
We have 3 approvals on this and it's been open for months. 
@open-telemetry/specs-semconv-approvers @open-telemetry/specs-semconv-maintainers is anything else holding this up? It's pretty small and entirely new experimental spec, required to make some traction on client instrumentation. Thanks!

File: docs/general/session.md
Comment (yurishkuro) [reply docs/general/session.md:7]: sorry if I was unclear - I think the explanation needs to be in the text, including the "who decides" part.

**Review by AlexanderWert — APPROVED** [review]

File: docs/general/session.md
Comment (bryce-b) [reply docs/general/session.md:7]: Ideally, we would provide a spec for this behavior, however it may be determined by then instrumentation vendor.

Comment (breedx-splk) [reply docs/general/session.md:7]: @yurishkuro would you mind moving this to an issue for future spec work? I think that could help us get to resolution on this PR. 🙏🏻

**Review by yurishkuro — APPROVED** [review]

File: docs/client/session.md
Comment (joaopgrassi) [reply docs/client/session.md]: >  want to know how many dropped/slow-render frames there within a given sessions to identify its health (or whatever). We would need this association for that to happen in a reasonable way.

This is pretty much what a exemplar is https://opentelemetry.io/docs/specs/otel/metrics/data-model/#exemplars.

If `session.id` is recorded on a span, and you have a metric, say `frames.render_time` with an exemplar, you could see a spike in render time, navigate to the trace and there you can see the session id. 

But one thing I'm not understanding, maybe I'm lacking context: What would knowing the sessionId help in such case (slow/dropped frames)? Wouldn't simply knowing that there's a slow down in frame rendering be enough to know there's a problem with the application? I would think, it would be valuable to know maybe which UI/Page it happened, which should be low cardinality and could be added to metrics. Sorry if I'm totally wrong here.

File: docs/client/session.md
Comment (joaopgrassi) [reply docs/client/session.md]: I agree, with @breedx-splk , `session.id` is a high cardinality field. I don't think it is a good idea to ever have that as metric attributes.

File: docs/general/session.md
Comment (joaopgrassi) [reply docs/general/session.md:7]: I agree this should probably be part of the spec. Also the part about "This value should be generated by the SDK and not supplied by the client", seems also like a impl detail and should be part of the SDK spec.

**Review by joaopgrassi — COMMENTED** [review]
Usually in conventions, we also describe things like span names, maybe a structure of a trace (in this case modeling a session). Is this something the group discussed and want to work on? Or the idea is to just define the attribute for now? 

I'm missing a bit more of context on why this change is needed. Is there any issue/meeting notes/discussion documented that could help understanding the need for this attribute? 

Also please add a changelog entry.

**Comment by breedx-splk** [comment]
@joaopgrassi respectfully, this has been discussed at length in the client SIG. Sessions don't really have a data shape themselves (or at least not yet in our mental model), it's more like a sequence of telemetry all grouped together, as identified by a session identifier. Hence this PR.

The idea is to get something started so that we can iterate on it. This is not intended to be comprehensive yet.

**Comment by joaopgrassi** [comment]
@bryce-b all good, the concept of a session is clear to me. I was more interested in for example an issue where the SIG discussions and agreements/plans are documented. I don't participate in the SIG so having that piece of extra context IMO helps the general semconv approvers.

**Review by joaopgrassi — APPROVED** [review]

**Review by martinkuba — APPROVED** [review]

**Comment by breedx-splk** [comment]
@opentelemetrybot rerun

**Comment by joaopgrassi** [comment]
This has several approvals and has been open for a while with no objections. Merging :)

**Pull request merged**
